design/open/lakefs-on-kv-testing-plan.md

```diff
@@ -0,0 +1,133 @@
+
+### ```pkg/diagnostics```
+* Only read access
+* No test coverage. Need to verify all is working the same after migration
+
+### ```pkg/graveler/ref```
+* Branch locking (read)
+* Repos, branches, commits and tags - read/write
+* 95.5% testing coverage
+* No performance/benchmarking
+* Missing:
+  * Migration tests
+  * Comparison benchmark
+
+### ```pkg/graveler/retention```
+* No direct DB access. Accesses are done using ```pkg/graveler/ref```, which should make the DB transition seamless
+* 42.9% coverage in unit testing
+* Missing performance tests that might help detect degradation
+
+### ```pkg/graveler/staging```
+* 89.1% coverage
+* Covers all staged KV functionality
+* No performance/benchmarking
+* Missing:
+  * Migration tests
+  * Comparison benchmark
+
+### Ref-Dump/Restore
+Currently this functionality is not covered directly, but since it relies on '''pkg/graveler/ref''' for DB access, the DB migration should be seamless
+It can be leveraged, however, to extend the cover of '''pkg/graveler/ref''' and for performance, as it is quite exhaustive (traverses all branches and commits and tags, per repo)
+
+## Execution Plan
+
+### KVM1
+* Data level migration tests infrastructure
+  * Migrate data from Table to KV, extract both and compare
+    * Data extraction should be done by listing all objects in the DB, using a designated 'get' function, and compare
+  * Implement ~~dumpers for `gateway_multiprts`~~ `GetAll` for `multiparts.Tracker`, to return a list of `MultipartUpload`
+  * Implement comparison of `MultipartUploads` list. Lists are considered identical if objects are identical, but **not necessarily** at the same order
+  * Implement unit tests for `pkg/gateway/multiparts`
+    * Add multipart uploads using `multiparts.Tracker.Create` with Table DB (KV Feature Flag off)
+      * Create an entry with key representing each off the supported storages:
+        * azure, google, s3, local, mem & transient
+      * Read all entries using `GetAll` above (Table DB)
+      * Run migration for `gateway_multiparts` table
+      * Read all entries using `GetAll` (KV Store)
+      * Compare the lists and expect equality (up to order)
+* Infrastructure for running migration during a system test execution
+  * System test to run migration during multiparts upload
+  * Currently there is a single simple multiparts system test (single file, 7 parts) - this is also an opportunity to expand that
+* KV Store unit tests 
+* `multiparts.Tracker` benchmark to run on both Table DB and KV Store (use feature flag to toggle) and verify there is no degradation
+  * Define sequence(s) of actions to perform (Create/Get/Delete etc.)
+  * Run each sequence with feature flag off and on
+  * Compare results and fail if KV performance is more than [TBD]% slower
+  * Need an infrastructure 
+
+### Next MS
+TBD
+
+
+## Open questions
+* What performance degradation is acceptable? (Can be 0)
+* How do we simulate KV Store failure? Do we need it at all?
```

Comment:
We can create a mock for either Store or StoreMessage to simulate failures.
I would describe what kind of fault injection do we need to support to better understand the faults requirements

Reply:
Sounds good to me. Thanks

Reply:
Added to the doc